xproc/src/main/xml/specification.xml

```diff
@@ -2812,31 +2795,52 @@ they are <firstterm>visible</firstterm> to each other.
       <para><error code="S0036">All the step types in a pipeline or library <rfc2119>must</rfc2119>
           have unique names: it is a <glossterm>static error</glossterm> if any step type name is
           built-in and/or declared or defined more than once in the same scope.</error></para>
-
-<para>The scope of the names of the steps themselves is determined by
+</section>
+      
+<section>
+  <title>Scoping of step names</title>
+<para>The scope of the names of the steps (the values of the step's <code>name</code> attributes) is determined by
 the <glossterm>environment</glossterm> of each step. In general, the
 name of a step, the names of its sibling steps, the names of any steps
 that it contains directly, the names of its ancestors, and the names
 of the siblings of its ancestors are all in a common scope. <error
 code="S0002">All steps in the same scope <rfc2119>must</rfc2119> have
 unique names: it is a <glossterm>static error</glossterm> if two steps
 with the same name appear in the same scope.</error></para>
-
+</section>
+<section>
+  <title>Scoping of port names</title>
 <para>The scope of an input or output port name is the step on which
 it is defined. The names of all the ports on any step
 <rfc2119>must</rfc2119> be unique.</para>
-
-<para>Taken together, these uniqueness constraints guarantee that the
+      
+<para>Taken together with the scoping of step names, these uniqueness constraints guarantee that the
 combination of a step name and a port name uniquely identifies exactly
 one port on exactly one in-scope step.</para>
-
-<para>The scope of option and variable names is determined by where
-they are declared. When an option is declared with <tag>p:option</tag>
-(or a variable with <tag>p:variable</tag>), unless otherwise
-specified, its scope consists of the sibling elements that follow its
+</section>
+      
+<section>
+  <title>Scoping of non-static options and variables</title>
+<para>The scope of non-static option and variable names is determined by where
+they are declared. Their scope consists of the sibling elements that follow its
 declaration and the descendants of those siblings.
 </para>
+<para>Non-static options and variables declared in parent steps are not visible in child steps.</para>
```

Comment:
Maybe make it a bit more precise: “parent step declarations” and “child step declarations”. If that’s what you mean. 

Otherwise one could ask oneself whether `$foo` isn’t visible in `p:for-each` in this scenario:

```xml
<p:viewport match="something" name="parent">
  <p:variable name="foo" select="/something/bar"/>
  <p:for-each name="child">
    <p:variable name="$baz" select="count($foo)"/>
    …
  </p:for-each>
</p:viewport>
```

The viewport is the parent step, and the for-each is the child step. `$foo` is clearly in scope in the child step. In order not to be misleading, you need to speak of parent and child declarations, I guess.

Reply:
Thanks, better still. Will change.